### ISSUE TITLE
[Phase 3] U-Net 1D Model

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/speech-enhancement-system#4]
## Phase 3: U-Net 1D Model

**Parent:** #1
**Depends on:** #2

### Objectives
Implement U-Net architecture for waveform-domain speech enhancement.

### Tasks

- [ ] Implement encoder blocks with downsampling
- [ ] Implement decoder blocks with upsampling
- [ ] Add skip connections between encoder and decoder
- [ ] Handle variable input lengths
- [ ] Test end-to-end forward pass
- [ ] Verify output shape matches input

### Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/models.py` | Update | Add UNet1D class |

### Code Reference

```python
class UNet1D(nn.Module):
    def __init__(self, in_channels: int = 1):
        super().__init__()

        # Encoder
        self.enc1 = self._enc_block(in_channels, 64, 15)
        self.enc2 = self._enc_block(64, 128, 5)
        self.enc3 = self._enc_block(128, 256, 3)
        self.enc4 = self._enc_block(256, 512, 3)

        # Bottleneck
        self.bottleneck = nn.Sequential(
            nn.Conv1d(512, 1024, 3, padding=1),
            nn.BatchNorm1d(1024),
            nn.LeakyReLU(0.2)
        )

        # Decoder
        self.dec4 = self._dec_block(1024, 512, 3)
        self.dec3 = self._dec_block(1024, 256, 3)  # 512+512
        self.dec2 = self._dec_block(512, 128, 5)   # 256+256
        self.dec1 = self._dec_block(256, 64, 15)   # 128+128

        self.output = nn.Conv1d(128, 1, 1)  # 64+64

    def _enc_block(self, in_ch, out_ch, k):
        return nn.Sequential(
            nn.Conv1d(in_ch, out_ch, k, padding=k//2),
            nn.BatchNorm1d(out_ch),
            nn.LeakyReLU(0.2),
            nn.MaxPool1d(2)
        )

    def _dec_block(self, in_ch, out_ch, k):
        return nn.Sequential(
            nn.ConvTranspose1d(in_ch, out_ch, 2, stride=2),
            nn.BatchNorm1d(out_ch),
            nn.LeakyReLU(0.2)
        )

    def forward(self, x):
        # Encoder
        e1 = self.enc1(x)
        e2 = self.enc2(e1)
        e3 = self.enc3(e2)
        e4 = self.enc4(e3)

        b = self.bottleneck(e4)

        # Decoder with skip connections
        d4 = self.dec4(b)
        d3 = self.dec3(torch.cat([d4, e4], dim=1))
        d2 = self.dec2(torch.cat([d3, e3], dim=1))
        d1 = self.dec1(torch.cat([d2, e2], dim=1))

        return self.output(torch.cat([d1, e1], dim=1))
```

### Definition of Done

- [ ] Model forward pass works
- [ ] Skip connections properly concatenate features
- [ ] Output has same length as input (or predictable relationship)
- [ ] Handles batch dimension correctly
- [ ] No dimension mismatch errors

### Technical Notes

**For junior developers:**
- Skip connections help preserve fine details
- Use LeakyReLU to avoid dead neurons
- ConvTranspose1d upsamples by stride factor
- Padding must be set correctly to maintain dimensions
